maven-core/src/main/java/org/apache/maven/lifecycle/internal/MojoExecutor.java

```diff
@@ -135,21 +140,27 @@ else if ( Artifact.SCOPE_TEST.equals( classpath ) )
         return Collections.unmodifiableCollection( scopes );
     }
 
-    public void execute( MavenSession session, List<MojoExecution> mojoExecutions, ProjectIndex projectIndex )
+    public void execute( final MavenSession session,
+                         final List<MojoExecution> mojoExecutions,
+                         final ProjectIndex projectIndex )
         throws LifecycleExecutionException
 
     {
-        DependencyContext dependencyContext = newDependencyContext( session, mojoExecutions );
+        final DependencyContext dependencyContext = newDependencyContext( session, mojoExecutions );
 
-        PhaseRecorder phaseRecorder = new PhaseRecorder( session.getCurrentProject() );
+        final PhaseRecorder phaseRecorder = new PhaseRecorder( session.getCurrentProject() );
 
-        for ( MojoExecution mojoExecution : mojoExecutions )
+        mojosExecutionStrategy.execute( mojoExecutions, session, new MojoExecutionRunner()
         {
-            execute( session, mojoExecution, projectIndex, dependencyContext, phaseRecorder );
-        }
+            @Override
+            public void run( MojoExecution mojoExecution ) throws LifecycleExecutionException
+            {
+                MojoExecutor.this.execute( session, mojoExecution, projectIndex, dependencyContext, phaseRecorder );
+            }
+        } );
     }
 
-    public void execute( MavenSession session, MojoExecution mojoExecution, ProjectIndex projectIndex,
+    private void execute( MavenSession session, MojoExecution mojoExecution, ProjectIndex projectIndex,
```

Comment:
Well, this is now reduced in scope. Is this going to be a problem? It's a public component after all.

Reply:
We can revert this change, I don't think it will cause any problem to leave this method `public` for 3.x.
I think the reasoning for this change is that two of the parameters `DependencyContext` and `PhaseRecorder` are in an _internal_ package, thus not exposed to plugins anyway.

Reply:
I would propose to leave this as `public` and add Javadoc that the visiblity will change in 4.0.0.

Reply:
Then it should be flagged with `@Deprecated` in this branch ... ?

Reply:
Well, the method isn't to go away, will it? It will just change scope. I don't know whether `@Deprecated` is appropriate.

Reply:
From a user point of view, the method _is_ going away.  I would deprecate it in 3.8.x and make it private in 4.x.  That way, people are aware they should not use it anymore on 3.8.x.

Reply:
Agreed, please do so.